File: java/client/src/org/openqa/selenium/devtools/Command.java
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.openqa.selenium.devtools;

import com.google.common.collect.ImmutableMap;

import org.openqa.selenium.json.JsonInput;

import java.lang.reflect.Type;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;

public class Command<X> {

  private final String method;
  private final Map<String, Object> params;
  private final Function<JsonInput, X> mapper;
  private final boolean sendsResponse;

  public Command(String method, Map<String, Object> params) {
    this(method, params, Void.class);
  }

  public Command(String method, Map<String, Object> params, Type typeOfX) {
    this(method, params, input -> input.read(typeOfX));
    Objects.requireNonNull(typeOfX, "Type to convert to must be set.");
  }

  public Command(String method, Map<String, Object> params, Function<JsonInput, X> mapper) {
    this(method, params, mapper, true);
  }

  private Command(String method, Map<String, Object> params, Function<JsonInput, X> mapper, boolean sendsResponse) {
    this.method = Objects.requireNonNull(method, "Method name must be set.");
    this.params = ImmutableMap.copyOf(Objects.requireNonNull(params, "Command parameters must be set."));
    this.mapper = Objects.requireNonNull(mapper, "Mapper for result must be set.");

    this.sendsResponse = sendsResponse;
  }

  public String getMethod() {
    return method;
  }

  public Map<String, Object> getParams() {
    return params;
  }

  Function<JsonInput, X> getMapper() {
    return mapper;
  }

  public boolean getSendsResponse() {
    return sendsResponse;
  }

  /**
   * Some CDP commands do not appear to send responses, and so are really hard
   * to deal with. Work around that by flagging those commands.
   */
  public Command<X> doesNotSendResponse() {
    return new Command<>(method, params, mapper, false);
  }
}

File: dotnet/src/webdriver/BiDi/Command.cs
// <copyright file="Command.cs" company="Selenium Committers">
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
// </copyright>

using System.ComponentModel;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.Json.Serialization.Metadata;

namespace OpenQA.Selenium.BiDi;

public readonly record struct Command<TParameters, TResult>(
    string Method,
    JsonTypeInfo<TParameters> ParamsTypeInfo,
    JsonTypeInfo<TResult> ResultTypeInfo)
    where TParameters : Parameters
    where TResult : EmptyResult;

public record Parameters
{
    public static Parameters Empty { get; } = new Parameters();

    [JsonExtensionData]
    [EditorBrowsable(EditorBrowsableState.Never)]
    public Dictionary<string, JsonElement>? RawAdditionalData { get; internal set; }

    [JsonIgnore]
    public AdditionalData AdditionalData
    {
        get => RawAdditionalData is null ? AdditionalData.Empty : AdditionalData.FromDictionary(RawAdditionalData);
        init
        {
            if (value.IsEmpty)
            {
                RawAdditionalData = null;
            }
            else
            {
                RawAdditionalData = [];
                foreach (var prop in value)
                {
                    RawAdditionalData[prop.Name] = prop.Value;
                }
            }
        }
    }
}

public abstract record CommandOptions
{
    public TimeSpan? Timeout { get; init; }

    public AdditionalData AdditionalData { get; init; }

    public AdditionalData AdditionalMessageData { get; init; }
}

public abstract record EmptyResult
{
    [JsonExtensionData]
    [EditorBrowsable(EditorBrowsableState.Never)]
    // IMPORTANT NOTE: Reasons for this property being public:
    // - The setter is not internal because the deserializer needs to be able to set this property when deserializing a command result
    // - The property is public to make external serializers see it so that they can deserialize additional data when deserializing a command result
    // - EditorBrowsableState.Never hides this property from IntelliSense to avoid confusion for users; it is technically a public property
    public Dictionary<string, JsonElement>? RawAdditionalData { get; set; }

    [JsonIgnore]
    public AdditionalData AdditionalData
    {
        get => RawAdditionalData is null ? AdditionalData.Empty : AdditionalData.FromDictionary(RawAdditionalData);
    }

    [JsonIgnore]
    public AdditionalData AdditionalMessageData { get; internal set; }
}
